### DISCUSSION TITLE
Shared Properties - how to have expression for one type + is it legacy?

### DISCUSSION THREAD

**Discussion by matanshukry** [rrousselGit/freezed#976] [Q&A] [answered]
I would like to have a shared property between different types, where 1 of them isn't a property but a known expression.

Example:
```
@Freezed()
class Car with _$Car {
  const factory Car.green({required String uniquePropForGreen, required WidgetBuilder builder}) = _GreenCar;
  const factory Car.black({required String uniquePropForBlack, required WidgetBuilder builder}) = _BlackCar;
  const factory Car.gold({required String uniquePropForBlack}) = _GoldCar;
}
```

The `builder` for gold should be defined somewhere else - not per instance.
e.g. let's assume it's always Text('I am a gold car!').

Usage (that I want):
```
Widget buildCarWidget(BuildContext context, Car car) {
  return car.builder(context);
}
```

How can I do that?

Note: I've also noticed the entire `Union types and Sealed classes` is marked as Legacy, but pattern matching doesn't give you shared properties as far as I can tell; so is it really legacy?

**Comment by rrousselGit** [comment, marked as answer]
"when/map" are legacy. Writing a union isn't

There is no way to define `builder` in a single place unfortunately.

**Reply by matanshukry** [reply]
@rrousselGit I recommend adding the part about unions in the docs btw, it's a bit confusing in there.
Would also add an example of replacing when/map with the dart 3 syntax - but that's because I think people are lazy and it's easier that way xd

And cool, thanks for answering both questions!

fyi: I ended up writing a method for the above.
e.g.
```
@Freezed()
class Car with _$Car {
  const Car._();

  const factory Car.green({required String uniquePropForGreen, required WidgetBuilder builder}) = _GreenCar;
  const factory Car.black({required String uniquePropForBlack, required WidgetBuilder builder}) = _BlackCar;
  const factory Car.gold({required String uniquePropForBlack}) = _GoldCar;
  
  WidgetBuilder get builder => map(green: (grenen) => green.builder, black: (black) => black.builder, gold: (gold) => _myGuildBuilder);
}
```

As mentioned though, the last map should be replaced.
